src/sse42/stage1.rs

```diff
@@ -30,41 +47,43 @@ fn fill_input(ptr: &[u8]) -> SimdInput {
     }
 }
 
-#[cfg_attr(not(feature = "no-inline"), inline(always))]
-unsafe fn check_utf8(
-    input: &SimdInput,
-    has_error: &mut __m128i,
-    previous: &mut AvxProcessedUtfBytes,
-) {
-    let highbit: __m128i = _mm_set1_epi8(static_cast_i8!(0x80u8));
-    if (_mm_testz_si128(_mm_or_si128(input.v0, input.v1), highbit)) == 1 {
-        // it is ascii, we just check continuation
-        *has_error = _mm_or_si128(
-            _mm_cmpgt_epi8(
-                previous.carried_continuations,
-                _mm_setr_epi8(9, 9, 9, 9, 9, 9, 9, 9, 9, 9, 9, 9, 9, 9, 9, 1),
-            ),
-            *has_error,
-        );
-    } else {
-        // it is not ascii so we have to do heavy work
-        *previous = avxcheck_utf8_bytes(input.v0, &previous, has_error);
-        *previous = avxcheck_utf8_bytes(input.v1, &previous, has_error);
+struct Utf8CheckingState {
+    has_error: __m128i,
+    previous: ProcessedUtfBytes,
+}
+
+impl Default for Utf8CheckingState {
+    #[cfg_attr(not(feature = "no-inline"), inline)]
+    fn default() -> Self {
+        Utf8CheckingState {
+            has_error: unsafe { _mm_setzero_si128() },
+            previous: ProcessedUtfBytes::default(),
+        }
     }
+}
+
+#[inline]
+fn is_utf8_status_ok(has_error: __m128i) -> bool {
```

Comment:
probably use the same `#[cfg_attr(not(feature = "no-inline"), inline(always))]` as in other functions, it allows to debug performance by disabling inlining and seeing where time is spend but has no ill effect otherwise.

Reply:
that's me being lazy